File: cslacs10/CSLA.Test/SimpleRoot.vb
<Serializable()> _
Public Class SimpleRoot
  Inherits BusinessBase

  Private mData As String = ""

  Public Property Data() As String
    Get
      Return mData
    End Get
    Set(ByVal Value As String)
      If mData <> Value Then
        mData = Value
        MarkDirty()
      End If
    End Set
  End Property

  <Serializable()> _
Private Class Criteria
    Public Data As String

    Public Sub New()
      Data = "<new>"
    End Sub

    Public Sub New(ByVal Data As String)
      Me.Data = Data
    End Sub
  End Class

  Public Shared Function NewSimpleRoot() As SimpleRoot
    Dim crit As Criteria = New Criteria
    Dim result As Object = DataPortal.Create(crit)
    Return DirectCast(result, SimpleRoot)
  End Function

  Public Shared Function GetSimpleRoot(ByVal Data As String) As SimpleRoot
    Return DirectCast(DataPortal.Fetch(New Criteria(Data)), SimpleRoot)
  End Function

  Public Shared Sub DeleteSimpleRoot(ByVal Data As String)
    DataPortal.Delete(New Criteria(Data))
  End Sub


  Private Sub New()
    ' prevent direct creation
  End Sub

  Protected Overrides Sub DataPortal_Create(ByVal Criteria As Object)
    Dim crit As Criteria = DirectCast(Criteria, Criteria)
    mData = crit.Data
    Session.Add("Root", "Created")
  End Sub

  Protected Overrides Sub DataPortal_Fetch(ByVal Criteria As Object)
    Dim crit As Criteria = DirectCast(Criteria, Criteria)
    mData = crit.Data
    MarkOld()
    Session.Add("Root", "Fetched")
  End Sub

  Protected Overrides Sub DataPortal_Update()

    If IsDeleted Then
      ' we would delete here
      Session.Add("Root", "Deleted")
      MarkNew()
    Else
      If IsNew Then
        ' we would insert here
        Session.Add("Root", "Inserted")

      Else
        ' we would update here
        Session.Add("Root", "Updated")
      End If
      MarkOld()
    End If
  End Sub

  Protected Overrides Sub DataPortal_Delete(ByVal Criteria As Object)
    ' we would delete here
    Session.Add("Root", "Deleted")
  End Sub

End Class

File: Source/tests/Csla.test/AppContext/SimpleRoot.cs
//-----------------------------------------------------------------------
// <copyright file="SimpleRoot.cs" company="Marimer LLC">
//     Copyright (c) Marimer LLC. All rights reserved.
//     Website: https://cslanet.com
// </copyright>
// <summary>The unique ID of this object</summary>
//-----------------------------------------------------------------------

namespace Csla.Test.AppContext
{
  [Serializable]
  class SimpleRoot : BusinessBase<SimpleRoot>
  {
    private string _Data = string.Empty;

    /// <summary>
    /// The unique ID of this object
    /// </summary>
    protected override object GetIdValue()
    {
      return _Data;
    }

    /// <summary>
    /// The data value for this object
    /// </summary>
    public string Data
    {
      get { return _Data; }
      set
      {
        if (!_Data.Equals(value))
        {
          _Data = value;
          MarkDirty();
        }
      }
    }

    /// <summary>
    /// Criteria for DataPortal overrides
    /// </summary>
    [Serializable]
    internal class Criteria
    {
      public const string DefaultData = "<new>";

      private string _Data = string.Empty;

      public string Data
      {
        get { return _Data; }
        set { _Data = value; }
      }

      public Criteria()
      {
        _Data = DefaultData;
      }

      public Criteria(string Data)
      {
        _Data = Data;
      }
    }

    /// <summary>
    /// Handles new DataPortal Create calls
    /// </summary>
    /// <param name="criteria"></param>
    private void DataPortal_Create(object criteria)
    {
      Criteria crit = criteria as Criteria;
      _Data = crit.Data;

      TestResults.Add("Root", "Created");
    }

    /// <summary>
    /// Handles DataPortal fetch calls
    /// </summary>
    /// <param name="criteria"></param>
    protected void DataPortal_Fetch(object criteria)
    {
      Criteria crit = criteria as Criteria;
      _Data = crit.Data;

      MarkOld();
      TestResults.Add("Root", "Fetched");
    }

    /// <summary>
    /// 
    /// </summary>
    [Update]
    protected void DataPortal_Update()
    {
      if (IsDeleted)
      {
        TestResults.Add("Root", "Deleted");
        MarkNew();
      }
      else
      {
        if (IsNew)
        {
          TestResults.Add("Root", "Inserted");
        }
        else TestResults.Add("Root", "Updated");

        MarkOld();
      }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="criteria"></param>
    [Delete]
    protected void DataPortal_Delete(object criteria)
    {
      TestResults.Add("Root", "Deleted");
    }
  }
}